[TAN-449] Add endpoints to fetch dynamic and snapshotted data by report graphs

alexander-cit opened the pull request:
Reasoning and how it works
https://www.notion.so/citizenlab/How-to-organize-the-code-and-data-around-graph-data-snapshots-for-public-reports-86da2a5e860c443080b808587ddd91ca

alexander-cit committed:
[TAN-449] Add endpoints to fetch dynamic and snapshotted data by report graphs

alexander-cit committed:
[TAN-449] Graph data units API MVP with integration tests

alexander-cit committed:
[TAN-449] Add props to users_by_gender query

alexander-cit committed:
[TAN-449] Move queries to separate classes

alexander-cit committed:
Merge branch 'master' into TAN-449-data-for-reports

alexander-cit committed:
[TAN-449] Make names same as on FE

luucvanderzee committed:
_

alexander-cit committed:
[TAN-449] Permit all props

luucvanderzee committed:
_

alexander-cit committed:
Merge branch 'master' into TAN-449-data-for-reports

alexander-cit committed:
Merge branch 'TAN-460-move-to-analytics' into TAN-449-data-for-reports

alexander-cit committed:
[TAN-449] Allow nil report name

alexander-cit committed:
[TAN-449] Process multiple queries in a separate class

alexander-cit committed:
[TAN-449] Fix graphs by using ActionController::Parameters in Analytics::Query

alexander-cit committed:
[TAN-449] Make ReactionsByTime query more similar to FE version

alexander-cit committed:
[TAN-449] Add usePublishedReactionsByTime hook

alexander-cit committed:
[TAN-449] Do not require both startAt/endAt

The same change was dont on the FE
https://github.com/CitizenLabDotCo/citizenlab/pull/6239/files#diff-4c3e4c9f2e5893d92cf6dcf5df43bbc023522962eea498ddc3842bfc3c64f419

alexander-cit committed:
[TAN-538] Make analytics code autoloadable. Fix specs

alexander-cit committed:
[TAN-449] Fix failing data units spec

NoMethodError:
  undefined method `empty?' for nil:NilClass
  ./engines/commercial/analytics/app/services/analytics/multiple_queries.rb:42:in `run'

alexander-cit committed:
[TAN-533] Make ReactionsByTime work with 3 data sources: Analytics, Live and Published data units

alexander-cit committed:
[TAN-449] Add TODO item to remove unused try

alexander-cit commented:
@jamesspeake The code still has some TODOs, it's not going to be merged in master yet. It would be just good to hear your opinion on the general approach. It's more or less the same as discussed in [this document](https://www.notion.so/citizenlab/How-to-organize-the-code-and-data-around-graph-data-snapshots-for-public-reports-86da2a5e860c443080b808587ddd91ca), with just one change. Instead of two controllers for live and published data with multiple actions for every graph, we now have just one controller with 2 actions. See my comments in the doc ([1](https://www.notion.so/How-to-organize-the-code-and-data-around-graph-data-snapshots-for-public-reports-86da2a5e860c443080b808587ddd91ca?d=2f926f08d7994aa9a6f8acecbad88c91&pvs=4#f7127c5a30ad4c17b285a0ddfc204aaf), [2](https://www.notion.so/How-to-organize-the-code-and-data-around-graph-data-snapshots-for-public-reports-86da2a5e860c443080b808587ddd91ca?d=b8244dae8ad04762a552fd16a8cd2bca&pvs=4#4ea9f0e37a3940309e2d3471de061a79)).

luucvanderzee committed:
Move graph data unit fetching logic to api folder

luucvanderzee committed:
Make reportId required

luucvanderzee committed:
Merge pull request #6426 from CitizenLabDotCo/move-graph-data-units-hooks

Move graph data unit fetching logic to api folder

alexander-cit committed:
Merge branch 'TAN-449-publish-graph-data-for-citizens' into TAN-449-data-for-reports

alexander-cit committed:
[TAN-449] Rename report_builder_report_id to report_id

alexander-cit committed:
[TAN-449] Add published_at to report serializer

alexander-cit committed:
Merge branch 'TAN-449-publish-graph-data-for-citizens' into TAN-449-data-for-reports

alexander-cit committed:
[TAN-449] Publish report data on save to avoid some inconsistencies

Inconsistencies:
- If an admin creates a report, publishes it, and then adds a new graph
to it without publishing it again, this new graph will be displayed
to citizens, but it'll be empty.
- If an admin removes a graph without publishing, this graph will not be
shown to citizens.

alexander-cit committed:
[TAN-449] Avoid unnecessary join to project dimension

alexander-cit committed:
[TAN-449] Properly transform params to keyword arguments

jamesspeake commented on back/engines/commercial/report_builder/app/models/report_builder/published_graph_data_unit.rb:
```diff
@@ -0,0 +1,26 @@
+#  fk_rails_...  (report_id => report_builder_reports.id)
+#
+module ReportBuilder
+  class PublishedGraphDataUnit < ::ApplicationRecord
```
Might be personal preference but not entirely sure about the name `GraphDataUnit` - but don't have anything that is amazingly better - `GraphDataQuery`? `GraphDataStore`? Think probably leave as it is - I'll get used to it!

jamesspeake commented on back/engines/commercial/report_builder/app/policies/report_builder/graph_data_unit_policy.rb:
```diff
@@ -0,0 +1,14 @@
+module ReportBuilder
+  class GraphDataUnitPolicy < ::ApplicationPolicy
+    def live?
+      admin? && active?
```
Can only admins create reports?

alexander-cit replied:
Yes for now. Soon it'll be available to moderators too, but we still need to make some decisions here https://github.com/CitizenLabDotCo/citizenlab/pull/6451

alexander-cit committed:
[TAN-449] Update constant name

jamesspeake commented on back/engines/commercial/report_builder/app/services/report_builder/queries/reactions_by_time.rb:
```diff
@@ -0,0 +1,43 @@
+# rubocop:disable Naming/VariableName
+module ReportBuilder
+  class Queries::ReactionsByTime < Queries::Base
+    def query(startAt: nil, endAt: nil, projectId: nil, **_other_props)
```
Why do we need to use camel case here? Is it because the analytics API uses this?

alexander-cit replied:
Initially, I wanted to preserve the format stored in craftjs_json, but now I see that snake case is less confusing. Fixed https://github.com/CitizenLabDotCo/citizenlab/pull/6309/commits/c7a571df86224fb2f8108cc29b662d6c7727d9cb#diff-0856afa5c73249411a81498cfba375e9610789ad7cee8f2dcc0a836bb9332f9dR14

jamesspeake commented on back/engines/commercial/report_builder/app/services/report_builder/query_repository.rb:
```diff
@@ -0,0 +1,30 @@
+    protected
+
+    def run_query(json_query)
+      # TODO: investigate why we need parameters and not HashWithIndifferentAccess
```
Assume it works regardless of this TODO?

alexander-cit replied:
Yes, it does. I just want to make the code simpler.

alexander-cit replied:
@jamesspeake I started to use `HashWithIndifferentAccess` instead of `ActionController::Parameters` in this commit https://github.com/CitizenLabDotCo/citizenlab/pull/6309/commits/cc5e1eb89c3b2cce91c03313c588ce424cf501ab. Let me know if it can cause any issues.

jamesspeake commented on back/engines/commercial/report_builder/app/services/report_builder/report_publisher.rb:
```diff
@@ -0,0 +1,34 @@
+    @report = report
+  end
+
+  # TODO: move all CraftJS related logic in one class
```
Assume this is coming later?

alexander-cit replied:
Probably not. It will only make things more complicated.

jamesspeake commented on back/engines/commercial/report_builder/app/services/report_builder/report_publisher.rb:
```diff
@@ -0,0 +1,34 @@
+  def publish
+    ReportBuilder::PublishedGraphDataUnit.where(report_id: @report.id).destroy_all
+
+    # TODO: change when we use multiple locales
```
Ditto

alexander-cit replied:
Yes, it comes after Sebi's and Iva's content builder changes.

jamesspeake commented on back/engines/commercial/report_builder/app/services/report_builder/report_publisher.rb:
```diff
@@ -0,0 +1,34 @@
+    ReportBuilder::PublishedGraphDataUnit.where(report_id: @report.id).destroy_all
+
+    # TODO: change when we use multiple locales
+    craftjs_json = @report.layout.craftjs_jsonmultiloc
```
Can spot a big dependency here on the work that Sebi is doing with the homepage craftjs_json. But hopefully you're already aware of that which is one of the reasons you're not doing multi-lingual yet

alexander-cit replied:
Yes, exactly.

jamesspeake commented on back/engines/commercial/report_builder/app/services/report_builder/report_publisher.rb:
```diff
@@ -0,0 +1,34 @@
+
+    nodes.each do |node_id, node_obj|
+      type = node_obj['type']
+      # TODO: is_a? is ugly! fix it
```
Why? Does what it needs to - although I'd generally use `instance_of?`

jamesspeake commented on back/spec/support/api_authentication_helper.rb:
Not entirely clear why you've changed this? I can see that header_token_for returns the same, but not sure why you need the additional method?

alexander-cit replied:
Because I used this `authorization_header` in my request spec. `header_token_for` doesn't work there.

jamesspeake commented on back/engines/commercial/report_builder/spec/requests/data_units_publishing_spec.rb:
Why is this a 'requests' test and not 'acceptance'? Doesn't seem to follow our usual testing pattern

alexander-cit replied:
Initially, it was an integration test that requested multiple endpoints in one spec (live data, publish, published data), and it's not possible with acceptance specs. But as now we got rid of publishing, indeed it doesn't make sense to keep it. So, I converted it into our usual acceptance spec https://github.com/CitizenLabDotCo/citizenlab/pull/6309/commits/cabac10dab623bfc75e89a5c96b98c09d7012499

jamesspeake reviewed (approved):
Generally looks pretty good to me. A few minor comments/questions here and there, though I do also wonder if we could do with a bit more test coverage - there don't seem to be many for a brand new model.

alexander-cit committed:
[TAN-449] Use correct project id column. Fix specs

alexander-cit committed:
[TAN-449] Pass controller data that we use

alexander-cit committed:
[TAN-449] Do not publish data for global reports

alexander-cit committed:
[TAN-449] Use snakecase in query keyword arguments

alexander-cit committed:
[TAN-449] Convert request spec into acceptance one

alexander-cit committed:
[TAN-449] Use HashWithIndifferentAccess in analytics Query

ActionController::Parameters was not properly used wiht `permit`.
ActionController::Parameters was only causing multiple hacks like `to_unsafe_h`
and clunky API when we needed to convert queries to
ActionController::Parameters.

alexander-cit committed:
[TAN-449] Use deep_dup in case query already has :page key

alexander-cit committed:
[TAN-549] Allow moderators to manage phase reports

alexander-cit committed:
[TAN-551] Add unit tests

alexander-cit committed:
[TAN-551] Improve specs

alexander-cit committed:
[TAN-551] Add policy specs

alexander-cit committed:
[TAN-551] Clean up report policy spec

alexander-cit committed:
[TAN-551] Allow moderators to request live graph data. Add specs

alexander-cit committed:
Merge pull request #6452 from CitizenLabDotCo/TAN-551-more-specs

[TAN-551] Add unit tests

alexander-cit committed:
Merge pull request #6451 from CitizenLabDotCo/TAN-549-permit-moderators-to-create-report

[TAN-549] Allow moderators to manage phase reports

luucvanderzee committed:
Get correct data in front office

alexander-cit committed:
[TAN-449] Skip necessary user authentication for published data
